### PR TITLE
Feature/alt mongo credentials

### PR DESCRIPTION
**Pull request by tzjames**
### Features and Changes
For some people, especially those who use AWS secrets manager to store secrets might find it easier to store separate env vars for composing their MONGODB_URI.  

https://growthbookusers.slack.com/archives/C01T6Q1SVFV/p1697697307803839

### Testing
Set MONGODB_URI env var.
Start app and see it load.
Remove MONGODB_URI env var and set the other env vars.
Restart the app and see it load.

### CODE REVIEW

File: packages/back-end/src/util/secrets.ts [hunk 1]
```diff
@@ -22,11 +22,27 @@ export const UPLOAD_METHOD = (() => {
   return "local";
 })();
 
-export let MONGODB_URI =
-  process.env.MONGODB_URI ??
-  (prod ? "" : "mongodb://root:password@localhost:27017/test?authSource=admin");
```
Comment (tzjames) [inline packages/back-end/src/util/secrets.ts:27]: Your code is missing this logic.  If it is still not defined after the alternate mongo db env vars AND not prod, then we want to use this as the default.

File: packages/back-end/src/util/secrets.ts [hunk 2]
```diff
@@ -22,11 +22,27 @@ export const UPLOAD_METHOD = (() => {
...
+let MONGODB_URI = process.env.MONGODB_URI;
+
+if (!MONGODB_URI) {
+  // Check for alternate mongo db environment variables
+  if (
+    process.env.MONGODB_USERNAME &&
+    process.env.MONGODB_PASSWORD &&
+    process.env.MONGODB_HOSTNAME
+  ) {
+    MONGODB_URI = `mongodb://${process.env.MONGODB_USERNAME}:${process.env.MONGODB_PASSWORD}@${process.env.MONGODB_HOSTNAME}/growthbook`;
```
Comment (tzjames) [inline packages/back-end/src/util/secrets.ts:34]: You might as well also add MONGODB_DBNAME and have it default to `growthbook` if it is not set.

File: packages/back-end/src/util/secrets.ts [hunk 3]
```diff
@@ -35,6 +51,10 @@ if (MONGODB_URI.match(/:27017(\/)?$/)) {
   MONGODB_URI = trimEnd(MONGODB_URI, "/") + "/test?authSource=admin";
 }
 
+export { MONGODB_URI };
+
+
+
```
Comment (tzjames) [inline packages/back-end/src/util/secrets.ts:57]: nit: a bit excessive on the new lines.

**Review by tzjames — COMMENTED** [review]
Thanks for doing this!  I left a few comments, some which need to be fixed to make our linter happy.

File: packages/back-end/src/util/secrets.ts (hunk 1 above)
Comment (yepher) [reply packages/back-end/src/util/secrets.ts:27]: That logic is moved down lower on line 50. Did I get that wrong?

File: packages/back-end/src/util/secrets.ts (hunk 2 above)
Comment (yepher) [reply packages/back-end/src/util/secrets.ts]: Good point. I added it.

File: packages/back-end/src/util/secrets.ts [hunk 4]
```diff
@@ -22,19 +22,38 @@ export const UPLOAD_METHOD = (() => {
...
+    // Add extra args if they exist
+    if (process.env.MONGODB_EXTRA_ARGS) {
+      MONGODB_URI += `?${process.env.MONGODB_EXTRA_ARGS}`;
+    }
+  }
 }
 
 // For backwards compatibility, if no dbname is explicitly set, use "test" and add the authSource db.
 // This matches the default behavior of the MongoDB driver 3.X, which changed when we updated to 4.X
-if (MONGODB_URI.match(/:27017(\/)?$/)) {
```
Comment (tzjames) [inline packages/back-end/src/util/secrets.ts:34]: This logic needs to remain for the case that MONGODB_URI = `mongodb://root:password@localhost:27017/`.  I would suggest the following logic before this if:
```
if(!MONOGODB_URI) {
   if (prod) {
      throw new Error("Missing MONGODB_URI or required alternate environment variables to generate it");
   }
  MONGODB_URI = "mongodb://root:password@localhost:27017/test?authSource=admin"
}
```

File: packages/back-end/src/util/secrets.ts [hunk 5]
```diff
@@ -22,19 +22,38 @@ export const UPLOAD_METHOD = (() => {
...
-  MONGODB_URI = trimEnd(MONGODB_URI, "/") + "/test?authSource=admin";
+if (!MONGODB_URI) {
+  MONGODB_URI = process.env.MONGODB_URI ?? (prod ? "" : "mongodb://root:password@localhost:27017/test?authSource=admin");
 }
 
+// Check if MONGODB_URI is still not set (either from environment or from the above generation)
+if (!MONGODB_URI) {
+  throw new Error("Missing MONGODB_URI or required environment variables to generate it");
+}
```
Comment (tzjames) [inline packages/back-end/src/util/secrets.ts:53]: This can then be gotten rid of.

**Review by tzjames — COMMENTED** [review]
Thank again for making those changes.  I think unfortunately there are a few more things to fix...

**Comment by tzjames** [comment]
Closing in favor of: https://github.com/growthbook/growthbook/pull/1772

**Pull request closed without merging**